File: go/problem189_rotateArray.go
// https://leetcode.com/problems/rotate-array

func rotate(nums []int, k int)  {
    size := len(nums)
    k = k%size

    if k <= 0 || size < 2 {
        return
    }

    // revert all
    for i,j := 0, size-1; i<j; i,j = i+1, j-1 {
        nums[i], nums[j] = nums[j], nums[i]
    }

    // revert request
    for i,j := 0, k-1; i<j; i,j = i+1, j-1 {
        nums[i], nums[j] = nums[j], nums[i]
    }

    // revert rest
    for i,j := k, size-1; i<j; i,j = i+1, j-1 {
        nums[i], nums[j] = nums[j], nums[i]
    }
    
}

File: java/Problem189_RotateArray.java
// https://leetcode.com/problems/rotate-array/

class Solution {
    public void rotate(int[] nums, int k) {
        int size = nums.length;

        if (k < 0 || size < 2 || size == k || k>(Math.pow(10, 5)) || size >(Math.pow(10, 5))) 
            return;

        while (k > size) {
            k = k-size;
        }
        
        int index = k-1;
        int[] copy = new int[size--];

        for (int i=0; i<=size; i++) {
            copy[i] = nums[i];
            if (index >= 0) {
                nums[i] = nums[size - index];
                index--;
            } else {
                nums[i] = copy[i - k];
            }
        }
    }
}
